compiler/rustc_typeck/src/check/coercion.rs

```diff
@@ -1448,7 +1448,15 @@ impl<'tcx, 'exprs, E: AsCoercionSite> CoerceMany<'tcx, 'exprs, E> {
                 expected.is_unit(),
                 pointing_at_return_type,
             ) {
-                if cond_expr.span.desugaring_kind().is_none() {
+                // If the block is from a macro, then do not suggest
+                // adding a semicolon, because there's nowhere to put it.
+                // See issue #81943.
+                let hir = fcx.tcx.hir();
+                let body_owner =
+                    hir.body_owner(hir.body_owned_by(hir.enclosing_body_owner(fcx.body_id)));
+                let from_same_crate = cond_expr.span.ctxt().dollar_crate_name()
+                    == hir.span(body_owner).ctxt().dollar_crate_name();
+                if cond_expr.span.desugaring_kind().is_none() && from_same_crate {
```

Comment:
I'm not sure what you are trying to do here.
The previously used `cond_expr.span.desugaring_kind().is_none()` works in the right direction, but uses an incorrect function (`desugaring_kind` only reports expansions from compiler desugarings, but not from macros).

The logic here should looks like
```rust
if is_macro_expanded(cond_expr.span) {
    // emit suggestions
}
```
, but I'm not sure how the most popular `is_macro_expanded` function is called now, @estebank probably knows.
There may be multiple because people adding new ones are often not aware of the previous ones.

There's `fn in_external_macro`, but it seems to only by lints right now.

Reply:
`in_external_macro` is exactly what I want. I didn’t just want to check if it’s a macro. I also want to make sure it’s in a separate crate, because that’s the case where we need to suppress this diagnostic.

Reply:
It's converted to `in_external_macro` now.